README.md
# terminalplot
[![Build Status](https://travis-ci.org/kressi/terminalplot.svg)](https://travis-ci.org/kressi/terminalplot)
[![PyPi Version](https://img.shields.io/pypi/v/terminalplot.svg)](https://pypi.python.org/pypi/terminalplot)

![Console plot](/plot.png)

## Installation
    pip install terminalplot

## Usage
### Command Line
Get size of terminal emulator (tty)
```
$ plot -s
> Rows: 25 Columns: 80
```

Plot some points
```
$ plot -x '-1 0 1 2 3 4' -y '0.1 0.2 0.23 0.234 0.24'
```

### API
Plotting a graph
```python
from terminalplot import plot
x = range(100)
y = [i**2 for i in x]
plot(x, y)
```

Get size of current terminal window
```python
from terminalplot import get_terminal_size
get_terminal_size()
```


CHANGES.md
# Changelog terminalplot

## Version 0.4.0 (2023-09-30)
- Move CI/CD to GitHub Actions
- Add black, ruff and mypy
- Move to codecov.io from coveralls.io
- Move to Python 3.11

## Version 0.3.0 (2019-12-12)
- Add support for Windows

## Version 0.2.5 (2015-05-26)
- Added tests
- Changed licence to GPL

## Version 0.2.3 (2015-05-10)
- Fixed issue with cli in python 3

## Version 0.2.2 (2015-04-27)
- Added travis

## Version 0.2.1 (2015-03-09)
- Added classifiers

## Version 0.2 (2015-03-09)
- Added command line interface

## Version 0.1 (2015-03-07)
- First release.


tests/terminalplot-test.py
import unittest
import terminalplot

class TestSequenceFunctions(unittest.TestCase):

    def test_get_size(self):
        assertIsInstance(terminalplot.get_terminal_size(), tuple)

    def test_plot(self):
        assertTrue(True)

if __name__ == '__main__':
    unittest.main()

tests/test_terminalplot.py
import unittest
import sys
from contextlib import contextmanager

if sys.version_info.major > 2:
    from io import StringIO
else:
    from StringIO import StringIO

import terminalplot
from terminalplot import command as cli

@contextmanager
def capture_sys_output():
    capture_out = StringIO()
    try:
        sys.stdout = capture_out
        yield capture_out
    finally:
        sys.stdout = sys.__stdout__

class TestTerminalPlot(unittest.TestCase):

    def test_get_size(self):
        terminal_size = terminalplot.get_terminal_size()
        self.assertTrue(type(terminal_size) is tuple)

    def test_plot(self):

        with capture_sys_output() as stdout:
            terminalplot.plot( x=[1,2,3], y=[1,2,3], rows=7, columns=3 )
        
        expected_output = ( "  *\n"
                            " * \n"
                            "*  \n"
                            "\nMin x: 1 Max x: 3 Min y: 1 Max y: 3\n" )

        self.assertEqual(stdout.getvalue(), expected_output)        

class TestCommand(unittest.TestCase):

    def test_list_floats(self):
        self.assertEqual(cli.list_floats("1 0.1 -0.3"), [1, 0.1, -0.3])

    def test_parser(self):
        parser = cli.make_parser()
        args   = parser.parse_args(['-x', '1 0.1 -0.3', '-y', '5 0 0'])
        self.assertEqual(args.x, [1, 0.1, -0.3])
        self.assertEqual(args.y, [5, 0, 0])
        self.assertFalse(args.terminal_size)


terminalplot/test/test_terminalplot.py
import sys
import unittest
from contextlib import contextmanager
from io import StringIO

import terminalplot
from terminalplot import command as cli


@contextmanager
def capture_sys_output():
    capture_out = StringIO()
    try:
        sys.stdout = capture_out
        yield capture_out
    finally:
        sys.stdout = sys.__stdout__


class TestTerminalPlot(unittest.TestCase):
    def test_get_size(self):
        rows, columns = terminalplot.get_terminal_size()
        self.assertIsInstance(rows, int)
        self.assertIsInstance(columns, int)

    def test_plot(self):
        with capture_sys_output() as stdout:
            terminalplot.plot(x=[1, 2, 3], y=[1, 2, 3], rows=7, columns=3)

        expected_output = (
            "  *\n" " * \n" "*  \n" "\nMin x: 1 Max x: 3 Min y: 1 Max y: 3\n"
        )

        self.assertEqual(stdout.getvalue(), expected_output)


class TestCommand(unittest.TestCase):
    def test_list_floats(self):
        self.assertEqual(cli.list_floats("1 0.1 -0.3"), [1, 0.1, -0.3])

    def test_parser(self):
        parser = cli.make_parser()
        args = parser.parse_args(["-x", "1 0.1 -0.3", "-y", "5 0 0"])
        self.assertEqual(args.x, [1, 0.1, -0.3])
        self.assertEqual(args.y, [5, 0, 0])
        self.assertFalse(args.terminal_size)


terminalplot/terminalplot.py
import shutil


def plot(x, y, rows=None, columns=None):
    """
    x, y list of values on x- and y-axis
    plot those values within canvas size (rows and columns)
    """
    def_row, def_col = get_terminal_size()
    rows = rows if rows else def_row
    columns = columns if columns else def_col
    # offset for caption
    rows -= 4

    # Scale points such that they fit on canvas
    x_scaled = scale(x, columns)
    y_scaled = scale(y, rows)

    # Create empty canvas
    canvas = [[" " for _ in range(columns)] for _ in range(rows)]

    # Add scaled points to canvas
    for ix, iy in zip(x_scaled, y_scaled):
        canvas[rows - iy - 1][ix] = "*"

    # Print rows of canvas
    for row in ["".join(row) for row in canvas]:
        print(row)

    # Print scale
    print(
        "".join(
            [
                "\nMin x: ",
                str(min(x)),
                " Max x: ",
                str(max(x)),
                " Min y: ",
                str(min(y)),
                " Max y: ",
                str(max(y)),
            ]
        )
    )


def scale(x, length):
    """
    Scale points in 'x', such that distance between
    max(x) and min(x) equals to 'length'. min(x)
    will be moved to 0.
    """
    s = (
        float(length - 1) / (max(x) - min(x))
        if x and max(x) - min(x) != 0
        else length
    )
    return [int((i - min(x)) * s) for i in x]


def get_terminal_size():
    return shutil.get_terminal_size()


terminalplot/command.py
"""
terminalplot.command.py
=======================

Implements command line interface for terminalplot.


usage: plot [-h] [-v] [-x X] [-y Y] [-s]

Minimalistic plot of a graph in terminal.

optional arguments:
  -h, --help           show this help message and exit
  -v, --version        show program's version number and exit
  -x X                 Values on x-axis. If empty, values on y-axis will be
                       plotted according to their indices, e.g. (0, y0),
                       (1, y1), ...
  -y Y                 Values on y-axis, at least one value has to be provided.
  -s, --terminal-size  Returns size of terminal <columns>, <rows>. No plot will
                       be generated.

Either -s or -x and -y have to be provided.

Example:
        plot -x '-1 0 1 2 3 4' -y '0.1 0.2 0.23 0.234 0.24'
        plot -s

"""

from argparse import ArgumentParser, RawTextHelpFormatter

from terminalplot import get_terminal_size, plot

from ._version import __version__


def list_floats(value):
    return [float(v) for v in value.split()]


def make_parser():
    parser = ArgumentParser(
        description="Minimalistic plot of a graph in terminal.",
        formatter_class=RawTextHelpFormatter,
        epilog=(
            "Either -s or -x and -y have to be provided.\n"
            "\nExample:\n"
            "\tplot -x '-1 0 1 2 3 4' -y '0.1 0.2 0.23 0.234 0.24'\n"
            "\tplot -s"
        ),
    )

    parser.add_argument(
        "-v", "--version", action="version", version="%(prog)s " + __version__
    )

    parser.add_argument(
        "-x",
        action="store",
        type=list_floats,
        help=(
            "Values on x-axis. If empty, values on y-axis will be plotted "
            "according to their indices, E.g. (0, y0), (1, y1), ..."
        ),
    )

    parser.add_argument(
        "-y",
        action="store",
        type=list_floats,
        help="Values on y-axis, at least one value has to be provided.",
    )

    parser.add_argument(
        "-s",
        "--terminal-size",
        action="store_true",
        help=(
            "Returns size of terminal <columns>, <rows>. No plot will be"
            "generated."
        ),
    )

    return parser


def main():
    parser = make_parser()
    args = parser.parse_args()

    if not args.terminal_size and not args.y:
        parser.error("Not enough arguments provided.")

    if args.y:
        y = args.y
        x = args.x or range(len(y))
        plot(x, y)

    if args.terminal_size:
        print("Rows: {}, Columns: {}".format(*get_terminal_size()))


if __name__ == "__main__":
    main()
